File: generator/src/main/java/com/navent/swagger/client/generator/ModelGenerator.java
package com.navent.swagger.client.generator;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.squareup.javapoet.*;
import org.json.JSONObject;

import javax.lang.model.element.Modifier;
import java.io.File;
import java.io.IOException;
import java.util.Date;
import java.util.Map;
import java.util.stream.Collectors;

public class ModelGenerator {


    public static void generate(String basePackage, String serviceName, String outputPath, JSONObject definition) {
        ModelGenerator.generateModels(definition.getJSONObject("definitions")).forEach(m -> {
            try {
                JavaFile.builder(basePackage + ".client.model", m)
                        .build()
                        .writeTo(new File(outputPath));
            } catch (IOException e) {
                e.printStackTrace();
            }
        });
    }

    static Iterable<TypeSpec> generateModels(JSONObject definitions) {
        return definitions.keySet().stream()
                .map(k -> generateModel(k, definitions.getJSONObject(k)))
                .collect(Collectors.toList());
    }

    public static TypeSpec generateModel(String name, JSONObject definition) {
        String type = definition.getString("type");

        TypeSpec.Builder builder;
        if ("object".equals(type) || null == type) {
            builder = generateObject(name, definition);
        } else {
            throw new UnknownModelTypeException(type);
        }

        return builder.addModifiers(Modifier.PUBLIC)
                .addAnnotation(lombok.Data.class)
                .build();
    }

    private static TypeSpec.Builder generateObject(String name, JSONObject definition) {
        return TypeSpec.classBuilder(name)
                .addFields(FieldGenerator.generateFields(definition.getJSONObject("properties")));
    }

    static class UnknownModelTypeException extends RuntimeException {
        UnknownModelTypeException(String message) {
            super(message);
        }
    }

    private static class FieldGenerator {
        static Iterable<FieldSpec> generateFields(JSONObject properties) {
            return properties.keySet().stream()
                    .map(k -> generateField(k, properties.getJSONObject(k)))
                    .collect(Collectors.toList());
        }

        static FieldSpec generateField(String name, JSONObject definition) {
            String type = definition.getString("type");

            FieldSpec.Builder builder;
            if ("integer".equals(type)) {
                builder = generateIntegerField(name, definition);
            } else if ("string".equals(type)) {
                builder = generateStringField(name, definition);
            } else if ("boolean".equals(type)) {
                builder = generateBooleanField(name, definition);
            } else if ("object".equals(type)) {
                builder = generateObjectField(name, definition);
            } else {
                throw new UnknownFieldTypeException(type);
            }

            if (definition.has("description")) {
                String description = definition.getString("description");

                builder.addJavadoc(description);
            }


            return builder.build();
        }

        private static FieldSpec.Builder generateIntegerField(String name, JSONObject definition) {
            String format = definition.getString("format");

            if ("int32".equals(format)) {
                return FieldSpec.builder(Integer.class, name, Modifier.PRIVATE);
            } else if ("int64".equals(format)) {
                return FieldSpec.builder(Long.class, name, Modifier.PRIVATE);
            } else {
                throw new UnknownIntegerFieldFormatException(format);
            }
        }

        private static FieldSpec.Builder generateBooleanField(String name, JSONObject definition) {
            return FieldSpec.builder(Boolean.class, name, Modifier.PRIVATE);
        }

        private static FieldSpec.Builder generateStringField(String name, JSONObject definition) {
            if (!definition.has("format")) {
                return FieldSpec.builder(String.class, name, Modifier.PRIVATE);
            } else {
                String format = definition.getString("format");

                if ("date-time".equals(format)) {
                    return FieldSpec.builder(Date.class, name, Modifier.PRIVATE)
                            .addAnnotation(AnnotationSpec.builder(JsonFormat.class)
                                    .addMember("shape", CodeBlock.builder().add("JsonFormat.Shape.STRING").build())
                                    .addMember("pattern", "$S", "dd-MM-yyyy hh:mm:ss")
                                    .build());
                } else {
                    throw new UnknownStringFieldFormatException(format);
                }
            }
        }

        private static FieldSpec.Builder generateObjectField(String name, JSONObject definition) {
            return FieldSpec.builder(ParameterizedTypeName.get(Map.class, String.class, Object.class), name, Modifier.PRIVATE);
        }

        static class UnknownFieldTypeException extends RuntimeException {
            UnknownFieldTypeException(String message) {
                super(message);
            }
        }

        static class UnknownIntegerFieldFormatException extends RuntimeException {
            UnknownIntegerFieldFormatException(String message) {
                super(message);
            }
        }

        static class UnknownStringFieldFormatException extends RuntimeException {
            UnknownStringFieldFormatException(String message) {
                super(message);
            }
        }
    }
}

File: generator/src/main/scala/com/navent/swagger/client/generator/ModelGenerator.scala
package com.navent.swagger.client.generator

import java.io.File
import javax.lang.model.element.Modifier

import com.navent.swagger.client.generator.Generator.Config
import com.squareup.javapoet._
import io.swagger.models._
import lombok.{Builder, Data}
import org.apache.commons.lang3.StringUtils

import scala.collection.JavaConverters._

object ModelGenerator {

  def generate(swagger: Swagger)(implicit config: Config): Iterable[TypeSpec] = {
    swagger.getDefinitions.asScala
      .map(p => generate(p._1, p._2))
  }

  def generate(name: String, model: Model)(implicit config: Config): TypeSpec = {
    val builder: TypeSpec.Builder = TypeSpec.classBuilder(name)
      .addModifiers(Modifier.PUBLIC)
      .addAnnotation(classOf[Data])
      .addAnnotation(classOf[Builder])

    val javadocBuilder = CodeBlock.builder()

    if(StringUtils.isNotBlank(model.getTitle))
      javadocBuilder.add("Title: $L\n", model.getTitle)

    if(StringUtils.isNotBlank(model.getDescription))
      javadocBuilder.add("$L\n", model.getDescription)

    if(model.getExample != null)
      javadocBuilder.add("Example: $L\n", model.getExample)

    builder.addJavadoc(javadocBuilder.build())

    PropertyGenerator.generate(model.getProperties.asScala).foreach(p => {
      builder.addField(p.field)
      builder.addTypes(p.types.asJava)
    })

    builder.build
  }

  private def writeToFile(t: TypeSpec)(implicit config: Config): Unit = {
    JavaFile.builder(config.modelPackage, t).build.writeTo(new File(config.codeOutput))
  }
}
